lib/tests/test_lexer.c

```diff
@@ -339,6 +345,38 @@ foo\n");
                "@version parsing mismatch, value %04x expected %04x", configuration->user_version, VERSION_VALUE_CURRENT);
 }
 
+Test(lexer, feature_flip_to_typing)
+{
+  gchar buf[128];
+  parser->lexer->ignore_pragma = FALSE;
+
+  start_grabbing_messages();
+
+  cfg_set_version_without_validation(configuration, 0);
+
+  gint flip_version = FEATURE_TYPING_MIN_VERSION - 1;
+  g_snprintf(buf, sizeof(buf), "@version: %d.%d\nbar\n",
+             (flip_version & 0xFF00) >> 8,
+             flip_version & 0xFF);
+  _input(buf);
+  assert_parser_identifier("bar");
+  cr_assert(configuration->user_version == flip_version);
+  assert_grabbed_log_contains("experimental behaviors of the future syslog-ng 4.0");
+
+  cr_assert(cfg_is_config_version_older(configuration, VERSION_VALUE_4_0));
+  cr_assert(cfg_is_typing_feature_enabled(configuration));
+
+  reset_grabbed_messages();
+
+  cfg_set_version_without_validation(configuration, 0);
+  _input("@version: 4.0\n\
```

Comment:
This one is also problematic when remapping versions.
Adding a new field for "4.0" (`VERSION_STR_4_0`) to versioning.h sounds like overkill, but can we, for example, just add a macro to the top of this test, so we can patch it internally?